### PR TITLE
[utils] Add MonoRefCount to factor ref-counting mechanisms

### PR DESCRIPTION
**Pull request by ludovic-henry**

### CODE REVIEW

**Comment by ludovic-henry** [comment]
cc @kumpera

File: mono/utils/mono-threads.c [hunk 1]
```diff
@@ -1151,15 +1151,15 @@ mono_threads_create_thread (MonoThreadStart start, gpointer arg, gsize * const s
 	MonoThreadHandle *ret;
 
 	thread_data = g_new0 (CreateThreadData, 1);
-	thread_data->ref = 2;
+	mono_refcount_init (thread_data, 1);
```
Comment (Therzok) [inline mono/utils/mono-threads.c]: Shouldn't this be 2?

Comment (Therzok) [inline mono/utils/mono-threads.c]: Nevermind, saw the inc below.

**Comment by kumpera** [comment]
❤️  this.

Though I believe mono_refcount_dec should take the destructor as an argument. Unchecked uses of _dec will lead to leaks.

File: mono/utils/mono-threads.c (hunk 1 above)
Comment (lambdageek) [inline mono/utils/mono-threads.c]: Why did this one change from 2 to 1?

**Comment by kumpera** [comment]
LGTM

It would be nice to follow up this PR with some docs on the header.

In particular, two things. 

- MonoRC must be the first field.

- Ref handoff requires external synchronization. By that I mean that inc/dec themselves only ensures correctness if there's external synchronization protecting the inc against a dec.

**Comment by ludovic-henry** [comment]
@kumpera I will add the doc right away.

I am split on whether to keep `mono_refcount_initialize` as a separate function, or only have `mono_refcount_new0`:
 - keep it would mean that you can have non-heap allocated memory, but when would that be needed? and do you really need a refcount mechanism for that?
 - remove it would mean that the destructor wouldn't free the memory, but we would do it in `mono_refcount_decrement` after calling the destructor; that would make the memory lifetime management fully handled at the same place.

**Pull request merged**